### DISCUSSION TITLE
Historical data

### DISCUSSION THREAD

**Discussion by romanmartushev** [weather-gov/api#270] [Q&A]
Im looking for record temperature highs and lows. Is this possible with your api? If not, any chance this could be added?

**Comment by tgherzog** [comment]
I'm looking for this too. I'm trying to send a query such as https://api.weather.gov/stations/KBWI/observations/2021-01-15T19:54:00+00:00 but it's returning 404

**Comment by netbymatt** [comment]
> I'm looking for this too. I'm trying to send a query such as https://api.weather.gov/stations/KBWI/observations/2021-01-15T19:54:00+00:00 but it's returning 404

There are several ins and outs to querying by date. Please see my post on stack overflow. https://stackoverflow.com/questions/50473592/api-weather-gov-start-and-end-times

**Reply by efishinsea** [reply]
Did any of you find good solution for historical data?  I'd like to simply query a given data for an area and get the 24 provided temps for that date so I can show High/Low. Currently, the hourly data (such as [https://api.weather.gov/gridpoints/PDT/142,62/forecast/hourly](url) ) will only show you temperatures from the current hour onward, so it can't be used to calculate or display the entire day.

**Reply by scadergit** [reply]
The hourly forecast does not show current temperatures, it shows the forecast (you'd think they would be similar, but they are not the same). To that end, you would never have a forecast for a previous hour, as you could get the actual observation.

**Reply by efishinsea** [reply]
good point!  is there another end point for historical data? I am not seeing that,,

**Reply by scadergit** [reply]
Historical data inclues actual observations taken by observation stations. You can see examples in the posts above yours. The world of observations is a bit more complex than a forecast, given different stations can do things a bit differently and record at different intervals. There is a not a single request for all data from a specific observation, you would need to make a request without the timestamp to determine what observations are available, then query for the observations timestamps you need.

**Reply by efishinsea** [reply]
Sorry, so I can easily view forecasted data for a location, but to be able to see the same 24 hour period temperatures for that same location requires multiple requests?  Maybe I'm explaining this incorrectly.   For a given date, I just want to know either the high/low temps for that date/place, or, the 24 "forecasted" or actual temps posted for that period.  Is there an example of this someplace?

I see an endpoint for https://api.weather.gov/stations/PDT/observations but am unable to get it to return any data.

**Reply by NateRL** [reply]
You might be able to have some luck using the /zones/forecast/{zoneId}/observations endpoint instead of the /stations/{stationId}/observations/{time} endpoint. No matter what I try, I have not been able to retrieve any data from /stations/{stationId}/observations/{time}. 

However, it was not too hard to get data from /zones/forecast/{zoneId}/observations. For example, this works fine for me:

https://api.weather.gov/zones/forecast/MNZ060/observations?start=2024-07-21T10%3A00%3A00Z&end=2024-07-21T11%3A00%3A00Z

The thing that is weird and frustrating is that when you request the zone-level observational data, you actually get station-level data in the return. So when I request the observations for zone MNZ060, the actual return data includes (for example) the observations from station KKMT (among others). However, I have not found a way to actually hit the stations endpoint to get the observational data from *just* KKMT. 

So if you want station-level data, you can probably get it by figuring out what zone it's in, hitting the endpoint for that entire zone, harvesting the station-level data that you want, and throwing the rest away. 

Not super efficient, but maybe that will work for you.

**Reply by sullynole** [reply]
Observations for a station such as KMKT: https://api.weather.gov/stations/KMKT/observations

https://api.weather.gov/stations/KMKT/observations/latest

**Reply by NateRL** [reply]
It's true that those endpoints work, but they don't help get *historical* data, which is the point of this thread. Unless I'm missing something? (I just checked, and it looks like /observations has ~1 week of data. That's not what I'd consider "historical".)

**Reply by sullynole** [reply]
> It's true that those endpoints work, but they don't help get *historical* data, which is the point of this thread. Unless I'm missing something? (I just checked, and it looks like /observations has ~1 week of data. That's not what I'd consider "historical".)

Correct, api.weather.gov does not contain observations beyond the last 7 days.  Historical data should be sourced from NCEI.

**Reply by NateRL** [reply]
Thanks so much! I didn't realize that api.weather.gov was limited to the last 7 days (although I'm not surprised). 

Shot in the dark, but do you have any pointers on reconciling historical observational data from NCEI with ongoing observational data from NWS? I'm building a machine learning model. I will train it on historical data (from NCEI, I guess) but then I need to monitor it in near-real time, so for that I'll need ongoing observational data (from NWS, I supposed). It would be really, really swell if I could find NCEI data and NWS data from the same observations so that the historical / observational data are actually the same. Any pointers?

**Reply by dwhitemv25** [reply]
Can you clarify "from the same observations"? The data is by station, so it sources from the same "data."

Note that the NCEI data goes through a QC process before becoming "official" and may vary from observed max/min.

**Reply by NateRL** [reply]
I haven't yet reconciled the stations between the NWS (weather.gov API) and NCEI sources, but I've already noticed that stations in the NWS data have 4-character identifiers and stations in the NCEI have 5-digit identifiers, so I wasn't sure if the data sources (the stations themselves) were the same or not. That's the next thing I'm working on.

The QC process is what I am going to look into after that, to see what the deviations look like between the raw NWS data and the QCd NCEI data. Depending on the variance (and how long it takes for the data to go through the QC process and show up in NCEI datasets), it may or may not be a problem. If I just have to wait, say, a week to get the QCd NCEI data that's what I'll do, and ignore the NWS observations (I still need their forecasts).

I'm still new to this whole area, and the data is really remarkably poorly documented, so I appreciate any of the help folks can offer.

Can you point me to a method for reconciling stations across NWS and NCEI data?

**Reply by dwhitemv25** [reply]
Sounds like you're trying to map between the MADIS identifiers and WMO/SYNOP identifiers. This might help (not sure how up to date it is) -- NCEI is having a bad day, its taking forever or failing to load pages, and several datasets are down, so can't do much more research on that right now:

https://weather.rap.ucar.edu/surface/stations.txt

The WMO/SYNOP ID can also be found in the Station Information on MesoWest. 

There's probably somewhere with a better mapping, but Google isn't helping much.

**Comment by mattdm** [comment]
See the Applied Climate Information System (ACIS) Web Services API, documented at https://www.rcc-acis.org/docs_webservices.html.
